### PR TITLE
Tab completion

### PR DESCRIPTION
**Pull request by hoelzro**
Do **NOT** merge me!  I'm not ready yet!

This adds the linenoise changes from the `no-linenoise` branch (and thus depends on the corresponding NQP and MoarVM changes), in addition to some rudimentary tab completion.

How rudimentary is it?  Read on...

  - [x] The list of completions is hardcoded to the list of classes, roles, and function in `OUTER::OUTER` from a Perl 6 script, because my Perl 6 introspection fu is weak.  Ideally it should get its list from the REPL's context.
  - [x] It relies on Linenoise.pm being installed, and that's not indicated in the README.  Maybe we should bundle it with Rakudo, or maybe just with Star.
  - [x] Linenoise.pm depends on linenoise being installed on your system; it doesn't bundle it.
  - [x] `linenoise` doesn't work out of the box on MoarVM, since MoarVM sets standard input to non-blocking.  I used a patched version of linenoise to get around this, but Linenoise.pm should probably do something about that.  Either that, or maybe MoarVM shouldn't set standard input to non-blocking; I don't know if that's a good idea or not.
  - [x] Completion is provided as a module; I don't know the best way to include the completion function in `Perl6::Compiler` otherwise, since the former is in Perl 6, and the latter is in NQP.
  - [ ] The way `Perl6::Compiler` loads Linenoise.pm and gets symbols from it is atrocious.
  - [ ] The completion function should also complete variables, attributes, and methods if possible.
  - [x] The completion function is coupled to linenoise; it should probably just return a `List` of completions.

### CODE REVIEW

**Comment by moritz** [comment]
sounds like a good proof of concept, but not mergable yet (due to Linenoise.pm/linenoise deps), but still, good work!

**Comment by hoelzro** [comment]
I should add that the REPL will still work *without* Linenoise.pm; tab completion just doesn't.

File: src/Perl6/Compiler.nqp [hunk 1]
```diff
@@ -93,4 +97,69 @@ class Perl6::Compiler is HLL::Compiler {
...
+            my @symbols := self.eval("use Linenoise; nqp::list(&linenoise, &linenoiseHistoryAdd, &linenoiseSetCompletionCallback, &linenoiseAddCompletion)");
+
+            $!linenoise := @symbols[0];
+            $!linenoise_add_history := @symbols[1];
+            my $linenoise_set_completion_callback := @symbols[2];
+            my $linenoise_add_completion := @symbols[3];
+
+            $!completions := nqp::hash();
+
+            my $core_keys := self.eval('CORE::.keys');
```
Comment (hoelzro) [inline src/Perl6/Compiler.nqp]: I don't care for this, but I don't know how else to get access to `CORE`.

File: src/Perl6/Compiler.nqp [hunk 2]
```diff
@@ -93,4 +97,69 @@ class Perl6::Compiler is HLL::Compiler {
         For more information, see the perl6(1) man page.\n"); 
         nqp::exit(0);
     }
+
+    method interactive(*@args, *%adverbs) {
```
Comment (hoelzro) [inline src/Perl6/Compiler.nqp]: I'm considering breaking this and perhaps `readline` out into a role/subclass just for interactive `Compiler`s.

**Comment by FROGGS** [comment]
I just merged nqp/master into https://github.com/perl6/nqp/compare/no-readlineint and nom into this branch, but get:

```
$ perl6-m
> 1
1
===SORRY!===
This type does not support elems
```

**Comment by FROGGS** [comment]
This seems to fix it, though @hoelzro can you take a look at merging both `.interactive` methods?

```diff
@@ -215,7 +214,7 @@ class Perl6::Compiler is HLL::Compiler {
                 if $m {
                     my $word := $m<word>;
                     unless $word ~~ /^ "&" <.upper>+ $/ {
-                        sorted_set_insert($!completions, $word);
+                        sorted_set_insert($!completions, $word) if $!completions;
                     }
                 }
             }
```

**Comment by hoelzro** [comment]
@FROGGS Thanks for looking at this; I'll fix this up right away!

**Comment by hoelzro** [comment]
@FROGGS I merged the interactive methods and fixed up the code to be based on nom.

**Comment by FROGGS** [comment]
```
$ perl6
> say 42
42
===SORRY!===
This type does not support elems
froggs@TL02L-R8RXTCW-linux:~/dev/rakudo$ perl6 --ll-exception
> say 42
42
This type does not support elems
   at src/Perl6/Compiler.nqp:7  (blib/Perl6/Compiler.moarvm:sorted_set_insert:10)
 from src/Perl6/Compiler.nqp:214  (blib/Perl6/Compiler.moarvm:readline:116)
 from gen/moar/stage2/NQPHLL.nqp:1202  (/home/froggs/dev/nqp/install/share/nqp/lib/NQPHLL.moarvm::26)
 from gen/moar/stage2/NQPHLL.nqp:1199  (/home/froggs/dev/nqp/install/share/nqp/lib/NQPHLL.moarvm:interactive:90)
 from src/Perl6/Compiler.nqp:151  (blib/Perl6/Compiler.moarvm:interactive:26)
 from gen/moar/stage2/NQPHLL.nqp:1383  (/home/froggs/dev/nqp/install/share/nqp/lib/NQPHLL.moarvm:command_eval:166)
 from src/Perl6/Compiler.nqp:59  (blib/Perl6/Compiler.moarvm:command_eval:93)
 from gen/moar/stage2/NQPHLL.nqp:1360  (/home/froggs/dev/nqp/install/share/nqp/lib/NQPHLL.moarvm:command_line:116)
 from src/gen/m-main.nqp:39  (/home/froggs/dev/nqp/install/share/perl6/runtime/perl6.moarvm:MAIN:18)
 from src/gen/m-main.nqp:35  (/home/froggs/dev/nqp/install/share/perl6/runtime/perl6.moarvm:<mainline>:197)
 from <unknown>:1  (/home/froggs/dev/nqp/install/share/perl6/runtime/perl6.moarvm:<main>:8)
 from <unknown>:1  (/home/froggs/dev/nqp/install/share/perl6/runtime/perl6.moarvm:<entry>:9)
froggs@TL02L-R8RXTCW-linux:~/dev/rakudo$ git show
commit 8dfe790d3f0dfc424d3685f5c6b1a3e734ba6b9a
Author: Rob Hoelz <rob@hoelz.ro>
Date:   Thu Apr 16 18:33:41 2015 -0500

    Return null string (instead of undefined) if linenoise is done
```

**Pull request closed without merging**